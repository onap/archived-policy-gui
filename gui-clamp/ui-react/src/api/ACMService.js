/*
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2021-2022 Nordix Foundation.
 *  ================================================================================
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  SPDX-License-Identifier: Apache-2.0
 *  ============LICENSE_END=========================================================
 */

export default class ACMService {

  static async getACMInstantiation() {

    const response = await fetch(window.location.pathname +
      'onap/policy/clamp/acm/v2/instantiation');

    return response
  }

  static async createInstanceProperties(instancePropertiesTemplate) {

    const response = await fetch(window.location.pathname +
      'onap/policy/clamp/acm/v2/instanceProperties', {
      method: 'POST',
      headers: {
        "Content-Type": "application/json"
      },
      credentials: 'same-origin',
      body: JSON.stringify(instancePropertiesTemplate),
    });

    return response
  }

  static async updateInstanceProperties(instanceName, version, instancePropertiesTemplate) {
    const params = {
      name: instanceName,
      version: version
    }

    const response = await fetch(window.location.pathname +
        'onap/policy/clamp/acm/v2/instanceProperties?' + (new URLSearchParams(params)), {
      method: 'PUT',
      headers: {
        "Content-Type": "application/json"
      },
      credentials: 'same-origin',
      body: JSON.stringify(instancePropertiesTemplate),
    });

    return response
  }

  static async deleteInstantiation(name, version) {
    const params = {
      name: name,
      version: version
    }

    const response = await fetch(window.location.pathname +
      'onap/policy/clamp/acm/v2/instanceProperties?' + (new URLSearchParams(params)), {
      method: 'DELETE',
      credentials: 'same-origin',
    });

    const data = await response;

    return data;
  }

  static async getInstanceOrderState(name, version) {
    const params = {
      name: name,
      version: version
    }
    const response = await fetch(window.location.pathname +
      'onap/policy/clamp/acm/v2/instantiationState'+ '?' + (new URLSearchParams(params)));

    const data = await response;

    return data;
  }

  static async changeInstanceOrderState(toscaObject) {
    const response = await fetch(window.location.pathname +
      'onap/policy/clamp/acm/v2/instantiation/command', {
      method: 'PUT',
      headers: {
        "Content-Type": "application/json"
      },
      credentials: 'same-origin',
      body: JSON.stringify(toscaObject)
    });

    return response
  }

  static async getToscaTemplate(name, version, instanceName) {
    const params = instanceName != null ?
      {
        name: name,
        version: version,
        instanceName: instanceName
      } :
      {
        name: name,
        version: version
      }

    const response = await fetch(window.location.pathname +
      'onap/policy/clamp/acm/v2/commission/toscaservicetemplate' + '?' + (new URLSearchParams(params)));

    const data = await response;

    return data;
  }

  static async uploadToscaFile(toscaObject) {
    const response = await fetch(window.location.pathname +
      'onap/policy/clamp/acm/v2/commission', {
      method: 'POST',
      headers: {
        "Content-Type": "application/json"
      },
      credentials: 'same-origin',
      body: JSON.stringify(toscaObject),
    });

    return response

  }

  static async deleteToscaTemplate(name, version) {
    const params = {
      name: name,
      version: version
    }

    const response = await fetch(window.location.pathname +
      'onap/policy/clamp/acm/v2/commission' + '?' + (new URLSearchParams(params)),
      {
        method: 'DELETE'
      });

    const data = await response;

    return data;
  }

  static async getCommonOrInstanceProperties(name, version, instanceName, isCommon) {
    const params = instanceName != null ?
      {
        name: name,
        version: version,
        instanceName: instanceName,
        common: isCommon
      } :
      {
        name: name,
        version: version,
        common: isCommon
      }

    const response = await fetch(window.location.pathname +
      'onap/policy/clamp/acm/v2/commission/getCommonOrInstanceProperties' + '?' + (new URLSearchParams(params)));

    return response;
  }

}
