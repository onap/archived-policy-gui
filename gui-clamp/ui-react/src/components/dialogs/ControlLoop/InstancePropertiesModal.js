/*
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2021 Nordix Foundation.
 *  ================================================================================
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  SPDX-License-Identifier: Apache-2.0
 *  ============LICENSE_END=========================================================
 */

import React, { useEffect, useState } from "react";
import Modal from "react-bootstrap/Modal";
import styled from "styled-components";
import Button from "react-bootstrap/Button";
import ControlLoopService from "../../../api/ControlLoopService";
import { JSONEditor } from "@json-editor/json-editor";
import Alert from "react-bootstrap/Alert";
import * as PropTypes from "prop-types";
import Form from "react-bootstrap/Form";
import Spinner from "react-bootstrap/Spinner";

const ModalStyled = styled(Modal)`
  @media (min-width: 800px) {
    .modal-xl {
      max-width: 96%;
    }
  }
  background-color: transparent;
`

const DivWhiteSpaceStyled = styled.div`
  overflow: auto;
  min-width: 100%;
  max-height: 300px;
  padding: 5px 5px 0px 5px;
  text-align: center;
`

const AlertStyled = styled(Alert)`
  margin-top: 10px;
`

const templateName = "ToscaServiceTemplateSimple";
const templateVersion = "1.0.0";
let tempJsonEditor = null;

function Fragment(props) {
  return null;
}

Fragment.propTypes = { children: PropTypes.node };
const InstancePropertiesModal = (props) => {
  const [show, setShow] = useState(true);
  const [toscaFullTemplate, setToscaFullTemplate] = useState({});
  const [jsonEditor, setJsonEditor] = useState(null);
  const [alertMessage, setAlertMessage] = useState(null);
  const [instancePropertiesGlobal, setInstancePropertiesGlobal] = useState({});
  const [serviceTemplateResponseOk, setServiceTemplateResponseOk] = useState(true);
  const [instancePropertiesResponseOk, setInstancePropertiesResponseOk] = useState(true);
  const [instanceName, setInstanceName] = useState('')
  const [isLoading, setIsLoading] = useState(true);

  useEffect(async () => {
    const toscaInstanceProperties = await ControlLoopService.getCommonOrInstanceProperties(templateName, templateVersion, false)
      .catch(error => error.message);

    const toscaTemplateResponse = await ControlLoopService.getToscaTemplate(templateName, templateVersion)
      .catch(error => error.message);

    if (!toscaInstanceProperties.ok) {
      const errorResponse = await toscaInstanceProperties.json()
      console.log(errorResponse)
      setInstancePropertiesGlobal(errorResponse);
      setInstancePropertiesResponseOk(false);
    }

    if (!toscaTemplateResponse.ok) {
      const errorResponse = await toscaTemplateResponse.json()
      console.log(errorResponse)
      setToscaFullTemplate(errorResponse)
      setServiceTemplateResponseOk(false);
    }

    if (toscaTemplateResponse.ok && toscaInstanceProperties.ok) {
      await parseJsonSchema(toscaTemplateResponse, toscaInstanceProperties);
    }

  }, []);

  const parseJsonSchema = async (fullTemplate, initialProperties) => {

    const fullJsonSchemaTemplate = await fullTemplate.json();
    setToscaFullTemplate(fullJsonSchemaTemplate);

    console.log(fullJsonSchemaTemplate);

    const filteredInitialStartValues = {};

    const instanceProperties = await initialProperties.json().then(properties => {
      const filteredTemplateObj = {};
      const propertiesTemplateArray = Object.entries(properties);

      propertiesTemplateArray.forEach(([key, value]) => {
        const propertiesObj = {
          properties: value.properties
        }

        const propValues = {};
        filteredTemplateObj[key] = propertiesObj;

        const jsonNodeSchemaKey = fullJsonSchemaTemplate.topology_template.node_templates[key]

        Object.entries(propertiesObj.properties).forEach(([pKey, pValue]) => {
          propValues[pKey] = jsonNodeSchemaKey.properties[pKey];
        });

        filteredInitialStartValues[key] = propValues;
      });

      return filteredTemplateObj;
    });

    const propertySchema = makeSchemaForInstanceProperties(instanceProperties);

    tempJsonEditor = createJsonEditor(propertySchema, filteredInitialStartValues);
    setJsonEditor(tempJsonEditor);
  }

  const makeSchemaForInstanceProperties = (instanceProps) => {
    const instancePropsArray = Object.entries(instanceProps);

    const newSchemaObject = {};

    newSchemaObject.title = "InstanceProperties";
    newSchemaObject.type = "object";
    newSchemaObject.properties = {};

    instancePropsArray.forEach(([key, value]) => {

      const propertiesObject = {};

      Object.entries(value.properties).forEach(([pKey, pValue]) => {
        propertiesObject[pKey] = {
          type: getType(pValue.type)
        }
      });

      newSchemaObject.properties[key] = {
        options: {
          "collapsed": true
        },
        properties: propertiesObject
      }
    });

    return newSchemaObject;
  }

  const getType = (pType) => {
    switch (pType) {
      case "map":
        return "string";
      case "string":
        return "string";
      case "integer":
        return "integer";
      case "list":
        return "array";
      case "object":
        return "object";
      default:
        return "object";

    }
  }

  const createJsonEditor = (fullSchema, instanceProperties) => {
    console.log(props.location.instanceName)
    setIsLoading(false)
    JSONEditor.defaults.options.collapse = true;

    return new JSONEditor(document.getElementById("editor"),
      {
        schema: fullSchema,
        startval: instanceProperties,
        theme: 'bootstrap4',
        iconlib: 'fontawesome5',
        object_layout: 'normal',
        disable_properties: false,
        disable_edit_json: false,
        disable_array_reorder: true,
        disable_array_delete_last_row: true,
        disable_array_delete_all_rows: false,
        array_controls_top: true,
        keep_oneof_values: false,
        collapsed: true,
        show_errors: 'always',
        display_required_only: false,
        show_opt_in: false,
        prompt_before_delete: true,
        required_by_default: false,
      });
  }

  const handleClose = () => {
    console.log('handleClose called');
    setShow(false);
    props.history.push('/');
  }

  const updateTemplate = (jsonEditorValues) => {
    const nodeTemplates = toscaFullTemplate.topology_template.node_templates;
    const instanceDataProperties = Object.entries(jsonEditorValues);

    instanceDataProperties.forEach(([key, value]) => {
      const nodeTemplatesKey = nodeTemplates[key]
      Object.entries(value).forEach(([pKey, pValue]) => {
        nodeTemplatesKey.properties[pKey] = pValue
      });
    });

    toscaFullTemplate.topology_template.node_templates = nodeTemplates;

    setToscaFullTemplate(toscaFullTemplate);

  }

  const handleSave = async () => {
    console.log("handleSave called")

    console.log("instanceName to be saved is: " + instanceName)

    updateTemplate(jsonEditor.getValue());

    const response = await ControlLoopService.createInstanceProperties(instanceName, toscaFullTemplate)
      .catch(error => error.message);

    if (response.ok) {
      successAlert();
    } else {
      await errorAlert(response);
    }
  }

  const handleNameChange = (e) => {
    setInstanceName(e.target.value)
  }

  const successAlert = () => {
    console.log("successAlert called");
    setAlertMessage(<Alert variant="success">
      <Alert.Heading>Instantiation Properties Success</Alert.Heading>
      <p>Instance Properties was successfully saved</p>
      <hr/>
    </Alert>);
  }

  const errorAlert = async (response) => {
    console.log("errorAlert called");
    setAlertMessage(<Alert variant="danger">
      <Alert.Heading>Instantiation Properties Failure</Alert.Heading>
      <p>An error occurred while trying to save</p>
      <p>Status code: { await response.status } : { response.statusText }</p>
      <p>Status Text: { await response.text() }</p>
      <hr/>
    </Alert>);
  }

  return (
    <ModalStyled size="xl"
                 show={ show }
                 onHide={ handleClose }
                 backdrop="static"
                 keyboard={ false }>
      <Modal.Header closeButton>
        <Modal.Title>Create Tosca Instance Properties</Modal.Title>
      </Modal.Header>
      <div style={ { padding: '5px 5px 0 5px' } }>
        <Modal.Body>
          <div id="editor"/>
          <AlertStyled show={ !serviceTemplateResponseOk }
                       variant="danger">Can't get service template:<br/>{ JSON.stringify(toscaFullTemplate, null, 2) }</AlertStyled>
          <AlertStyled show={ !instancePropertiesResponseOk }
                       variant="danger">Can't get instance properties:<br/>{ JSON.stringify(instancePropertiesGlobal, null, 2) }</AlertStyled>
        </Modal.Body>
        <DivWhiteSpaceStyled>
          { alertMessage }
        </DivWhiteSpaceStyled>
      </div>
      <Modal.Footer>
        <Button variant="primary" onClick={ handleSave }>Save</Button>
        <Button variant="secondary" onClick={ handleClose }>Close</Button>
      </Modal.Footer>
    </ModalStyled>
  );
}

export default InstancePropertiesModal;
