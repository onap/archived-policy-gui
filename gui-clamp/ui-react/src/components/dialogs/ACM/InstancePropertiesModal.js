/*
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2022 Nordix Foundation.
 *  ================================================================================
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  SPDX-License-Identifier: Apache-2.0
 *  ============LICENSE_END=========================================================
 */

import React, { useEffect, useState } from "react";
import Modal from "react-bootstrap/Modal";
import styled from "styled-components";
import Button from "react-bootstrap/Button";
import ACMService from "../../../api/ACMService";
import Alert from "react-bootstrap/Alert";
import * as PropTypes from "prop-types";
import InstantiationUtils from "./utils/InstantiationUtils";

const ModalStyled = styled(Modal)`
  @media (min-width: 800px) {
    .modal-xl {
      max-width: 96%;
    }
  }
  background-color: transparent;
`

const DivWhiteSpaceStyled = styled.div`
  overflow: auto;
  min-width: 100%;
  max-height: 300px;
  padding: 5px 5px 0px 5px;
  text-align: center;
`

const AlertStyled = styled(Alert)`
  margin-top: 10px;
`

const templateName = "ToscaServiceTemplateSimple";
const templateVersion = "1.0.0";

function Fragment(props) {
  return null;
}

Fragment.propTypes = { children: PropTypes.node };
const InstancePropertiesModal = (props) => {
  const [show, setShow] = useState(true);
  const [toscaFullTemplate, setToscaFullTemplate] = useState({});
  const [jsonEditor, setJsonEditor] = useState(null);
  const [alertMessage, setAlertMessage] = useState(null);
  const [instancePropertiesGlobal, setInstancePropertiesGlobal] = useState({});
  const [serviceTemplateResponseOk, setServiceTemplateResponseOk] = useState(true);
  const [instancePropertiesResponseOk, setInstancePropertiesResponseOk] = useState(true);
  const [instanceName, setInstanceName] = useState('')

  useEffect(async () => {
    const toscaTemplateResponse = await ACMService.getToscaTemplate(templateName, templateVersion)
        .catch(error => error.message);

    const toscaInstanceProperties = await ACMService.getCommonOrInstanceProperties(templateName, templateVersion, false)
      .catch(error => error.message);

    if (!toscaInstanceProperties.ok) {
      const errorResponse = await toscaInstanceProperties.json();
      console.log(errorResponse);
      setInstancePropertiesGlobal(errorResponse);
      setInstancePropertiesResponseOk(false);
    }

    if (!toscaTemplateResponse.ok) {
      const errorResponse = await toscaTemplateResponse.json();
      console.log(errorResponse);
      setToscaFullTemplate(errorResponse);
      setServiceTemplateResponseOk(false);
    }

    if (toscaTemplateResponse.ok && toscaInstanceProperties.ok) {
      const renderedJsonSchema = await InstantiationUtils.parseJsonSchema(toscaTemplateResponse, toscaInstanceProperties);
      setToscaFullTemplate(await (renderedJsonSchema).fullTemplate);
      setJsonEditor(await (renderedJsonSchema).jsonEditor);
    }

  }, []);

  const handleClose = () => {
    console.log('handleClose called');
    setShow(false);
    props.history.push('/');
  }

  const handleSave = async () => {
    console.log("handleSave called");

    setInstanceName(instanceName);

    console.log("instanceName to be saved is: " + instanceName);

    if (jsonEditor != null) {
      setToscaFullTemplate(InstantiationUtils.updateTemplate(jsonEditor.getValue(), toscaFullTemplate));
    }

    const response = await ACMService.createInstanceProperties(instanceName, toscaFullTemplate)
      .catch(error => error.message);

    if (response.ok) {
      successAlert();
    } else {
      await errorAlert(response);
    }
  }

  const successAlert = () => {
    console.log("successAlert called");
    setAlertMessage(<Alert variant="success">
      <Alert.Heading>Instantiation Properties Success</Alert.Heading>
      <p>Instance Properties was successfully saved</p>
      <hr/>
    </Alert>);
  }

  const errorAlert = async (response) => {
    console.log("errorAlert called");
    setAlertMessage(<Alert variant="danger">
      <Alert.Heading>Instantiation Properties Failure</Alert.Heading>
      <p>An error occurred while trying to save</p>
      <p>Status code: { await response.status } : { response.statusText }</p>
      <p>Status Text: { await response.text() }</p>
      <hr/>
    </Alert>);
  }

  return (
    <ModalStyled size="xl"
                 show={ show }
                 onHide={ handleClose }
                 backdrop="static"
                 keyboard={ false }>
      <Modal.Header closeButton>
        <Modal.Title>Create Tosca Instance Properties</Modal.Title>
      </Modal.Header>
      <div style={ { padding: '5px 5px 0 5px' } }>
        <Modal.Body>
          <div id="editor"/>
          <AlertStyled show={ !serviceTemplateResponseOk }
                       variant="danger">Can't get service template:<br/>{ JSON.stringify(toscaFullTemplate, null, 2) }</AlertStyled>
          <AlertStyled show={ !instancePropertiesResponseOk }
                       variant="danger">Can't get instance properties:<br/>{ JSON.stringify(instancePropertiesGlobal, null, 2) }</AlertStyled>
        </Modal.Body>
        <DivWhiteSpaceStyled>
          { alertMessage }
        </DivWhiteSpaceStyled>
      </div>
      <Modal.Footer>
        <Button variant="primary" onClick={ handleSave }>Save</Button>
        <Button variant="secondary" onClick={ handleClose }>Close</Button>
      </Modal.Footer>
    </ModalStyled>
  );
}

export default InstancePropertiesModal;
