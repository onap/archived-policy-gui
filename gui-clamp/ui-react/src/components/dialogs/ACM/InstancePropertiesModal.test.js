/*
 * ============LICENSE_START=======================================================
 *  Copyright (C) 2022 Nordix Foundation.
 *  ================================================================================
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  SPDX-License-Identifier: Apache-2.0
 *  ============LICENSE_END=========================================================
 */

import {shallow} from "enzyme";
import React from "react";
import InstancePropertiesModal from "./InstancePropertiesModal";
import toJson from "enzyme-to-json";
import { createMemoryHistory } from "history";
import { act } from "react-dom/test-utils";
import ACMService from "../../../api/ACMService";
import fullTemp from "./testFiles/fullTemplate.json";
import instanceProps from "./testFiles/instanceProps.json";

const instanceProperties = JSON.parse(JSON.stringify(instanceProps))
const fullTemplate = JSON.parse(JSON.stringify(fullTemp))
let logSpy = jest.spyOn(console, 'log');

describe('Verify InstancePropertiesModal', () => {

  const unmockedFetch = global.fetch
    beforeAll(() => {
      global.fetch = () =>
          Promise.resolve({
            status: 200,
            text: () => "OK",
            json: () => "{GlobalFetch}"
          });
    });

  afterAll(() => {
    global.fetch = unmockedFetch
  });

  beforeEach(() => {
    logSpy.mockClear()
  });

  it("renders without crashing", () => {
    shallow(<InstancePropertiesModal />);
  });

  it("renders correctly", () => {
    const tree = shallow(<InstancePropertiesModal />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should have save button element', () => {
    const container = shallow(<InstancePropertiesModal/>)
    expect(container.find('[variant="primary"]').length).toEqual(1);
  });

  it('should have close button element', () => {
    const container = shallow(<InstancePropertiesModal/>)
    expect(container.find('[variant="secondary"]').length).toEqual(1);
  });

  it('handleCreateUpdateToscaInstanceProperties called when save button clicked', () => {
    const component = shallow(<InstancePropertiesModal />);

    const instanceNameState = jest.spyOn(React, "useState");
    instanceNameState.mockImplementation(instanceName => [instanceName, 'Test']);

    act(() => {
      component.find('[variant="primary"]').simulate('click');
      expect(logSpy).toHaveBeenCalledWith('handleSave called');
    });
  });

  it('handleClose called when close button clicked', () => {
    const history = createMemoryHistory();
    const component = shallow(<InstancePropertiesModal history={ history }/>)

    act(() => {
      component.find('[variant="secondary"]').simulate('click');
      expect(logSpy).toHaveBeenCalledWith('handleClose called');
    });
  });

  it('handleSave called when save button clicked', () => {
    const component = shallow(<InstancePropertiesModal />)

    act(() => {
      component.find('[variant="primary"]').simulate('click');
      expect(logSpy).toHaveBeenCalledWith('handleSave called');
    });
  });

  it('Check useEffect is being called', async () => {
    const useEffect = jest.spyOn(React, "useEffect");
    const component = shallow(<InstancePropertiesModal />)
    act(async () => {
      expect(useEffect).toHaveBeenCalled();
      component.update()
    });
  });

  it('Check useEffect is being called', async () => {
    const useEffect = jest.spyOn(React, "useEffect");
    const component = shallow(<InstancePropertiesModal />)
    act(async () => {
      expect(useEffect).toHaveBeenCalled();
      component.update();
    });
  });

  it('getToscaTemplate gets called in useEffect with error',  async() => {
    const fetchMock = jest.spyOn(ACMService, 'getToscaTemplate').mockImplementation(() => Promise.resolve({
      ok: false,
      status: 200,
      text: () => "OK",
      json: () => fullTemplate
    }))

    const component = shallow(<InstancePropertiesModal/>)
    act(async () => {
      expect(fetchMock).toHaveBeenCalled();
      component.update();
    });
  });

  it('getCommonOrInstanceProperties gets called in useEffect with error',  async() => {
    const fetchMock = jest.spyOn(ACMService, 'getCommonOrInstanceProperties').mockImplementation(() => Promise.resolve({
      ok: false,
      status: 200,
      text: () => "OK",
      json: () => instanceProperties
    }))

    const component = shallow(<InstancePropertiesModal/>)
    act(async () => {
      expect(fetchMock).toHaveBeenCalled();
      component.update();
    });
  });

});
