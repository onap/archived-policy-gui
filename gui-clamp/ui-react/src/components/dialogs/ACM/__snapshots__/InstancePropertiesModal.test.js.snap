// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify InstancePropertiesModal renders correctly 1`] = `
<Styled(Modal)
  backdrop="static"
  keyboard={false}
  onHide={[Function]}
  show={true}
  size="xl"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle>
      Create Tosca Instance Properties
    </ModalTitle>
  </ModalHeader>
  <div
    style={
      Object {
        "padding": "5px 5px 0 5px",
      }
    }
  >
    <ModalBody>
      <FormGroup
        as={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "noGutters": false,
            },
            "render": [Function],
          }
        }
        controlId="formPlaintextEmail"
      >
        <FormLabel
          column={true}
          sm="2"
          srOnly={false}
        >
          Instance Name:
        </FormLabel>
        <input
          onChange={[Function]}
          sm="5"
          style={
            Object {
              "width": "50%",
            }
          }
          type="text"
          value=""
        />
      </FormGroup>
      <div
        id="editor"
      />
      <Styled(Alert)
        show={false}
        variant="danger"
      >
        Can't get service template:
        <br />
        {}
      </Styled(Alert)>
      <Styled(Alert)
        show={false}
        variant="danger"
      >
        Can't get instance properties:
        <br />
        {}
      </Styled(Alert)>
    </ModalBody>
    <styled.div />
  </div>
  <ModalFooter>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Save
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="secondary"
    >
      Close
    </Button>
  </ModalFooter>
</Styled(Modal)>
`;
